implementation

felicepagano opened the pull request.

felicepagano commented on src/main/java/it/fpagano/kata/goosegame/cell/Cell.java:
```diff
@@ -0,0 +1,41 @@
+  }
+
+  public Function<Player, Player> action(Function<Integer, Cell> f, Integer dadiResult, Player previewsState) {
+    return Function.identity();
```
ogni implementazione di cella ha la sua regola. tipo, vai avanti del numero ottenuto precedentemte dai dadi. oppure vai a una specifica casella. oppure nulla. Identity è non fare nulla, quindi restituisco lo stesso player. Viene restituito un nuovo player ogni qual volta eseguo un'azione, tranne quando non faccio nulla, quindi identity, questo mi permette anche di fare circuit breaking sulla funzione ricorsiva che uso per eseguire ricorsivamente le azioni delle celle sulle quali mi sto spostando.

marcodippy replied:
non sono troppo convinto che mi piaccia, il fatto che hai dovuto spiegare cosa faccia il metodo vuol dire che non è chiaro secondo me... anche qui vale il fatto che "cella" è dato e non logica, secondo me la logica non dovrebbe stare dentro `Cell`

felicepagano replied:
secondo me la cosa interessante di questo metodo è che se ci pensi alla fine non fa nulla se non restituire quello che dovresti fare una volta che il player è sulla cella. È lui poi che decide se farlo o meno. un po`come quando giochi ai giochi di società e vuoi barare. La cella ti dice cosa dovresti fare, ma sei tu player che poi applichi la regola o no.

felicepagano commented on src/main/java/it/fpagano/kata/goosegame/player/Player.java:
```diff
@@ -0,0 +1,82 @@
+    this.lastAction = lastAction;
+  }
+
+  public Tuple2<String, Integer> launch(List<Dado> dadi) {
```
rappresenta il messaggio generato da launch + il numero generato. mi porto dietro il logo.

felicepagano commented on src/main/java/it/fpagano/kata/goosegame/player/Player.java:
```diff
@@ -0,0 +1,82 @@
+    final List<Integer> dadiLanciati = dadi.map(Dado::generate);
+    final String messaggio = dadiLanciati.map(String::valueOf)
+        .collect(Collectors.joining(", ", this.name + " rolls ", "."));
+    return Tuple.of(messaggio, dadiLanciati.sum().intValue());
```
eh lancia i dadi, genera il messaggio e restituisce somma

felicepagano commented on src/main/java/it/fpagano/kata/goosegame/player/Player.java:
```diff
@@ -0,0 +1,82 @@
+    return Tuple.of(messaggio, dadiLanciati.sum().intValue());
+  }
+
+  public Player move(Tuple2<String, Integer> lancio, Function<Integer, Cell> f) {
```
si poi game dovrà comporre tutti questi pezzettini

felicepagano replied:
f è la funzione per generare le celle. in pratica non genero tutto il percorso ma ho una funzione che stabilisce cosa fanno le celle in determinate posizioni. quindi, passo l'indice della cella, creo l'istanza giusta della cella. cosi posso creare percorsi dinamici.

idx -> {
      if(idx > numeroDiCaselle) {
        return new Bounces(numeroDiCaselle);
      }
      switch (idx) {
        case 0: return new Start();
        case 6: return new Bridge(idx);
        case numeroDiCaselle: return new Win(idx);
        case 5:
        case 9:
        case 14:
        case 18:
        case 23:
        case 27: return new Goose(idx);
        default: return new Cell(idx);
      }
    };

marcodippy replied:
ma non c'è motivo di avere questa logica dentro l'oggetto `Player`

felicepagano commented on src/main/java/it/fpagano/kata/goosegame/Game.java:
```diff
@@ -0,0 +1,25 @@
+  private final Set<Player> players = new LinkedHashSet<>();
+
+  public String addPlayer(Player player) {
+    return players.add(player) ? playerAdded() : playerAlredyExists(player);
```
non lho capita

felicepagano replied:
devi leggerti il readme per capire perchè lo faccio. poi forse ci sta che non va bene fatto così

marcodippy replied:
boh a me non piace che torni messaggi in una stringa, per sapere se è andato a buon fine come fai?

felicepagano commented on src/main/java/it/fpagano/kata/goosegame/Game.java:
```diff
@@ -0,0 +1,25 @@
+
+public class Game {
+
+  private final Set<Player> players = new LinkedHashSet<>();
```
hashset di vavr è immutable. cosa devo cambiare?

marcodippy replied:
se hai "addPlayer" e non stai restituendo un nuovo `Game`, `Game` non è immutabile

marcodippy replied:
e cmq stai usando java.util.LinkedHashSet

felicepagano commented on src/main/java/it/fpagano/kata/goosegame/cell/Cell.java:
```diff
@@ -0,0 +1,41 @@
+
+public class Cell {
+
+  private final int idx;
```
non mi ci abitueró mai :D. ora lo faccio

felicepagano commented on src/main/java/it/fpagano/kata/goosegame/player/Player.java:
```diff
@@ -0,0 +1,82 @@
+public class Player {
+
+  private final String name;
+  private final Cell actualPosition;
```
non ho capito. actual devo rinomiarla?

marcodippy replied:
current = attuale
actual = effettivo

tu vuoi la `currentPosition` :)

felicepagano committed:
Init project

Initialize a maven project with maven-archetype-quickstart archetype.
Java compiler is 8. Dependecies: vavr

felicepagano committed:
Implement base case

felicepagano committed:
Update README file

felicepagano committed:
Introduce some javadoc

felicepagano committed:
Rename method to improve readability
